### PR TITLE
Filtering tuples of various types prior to iteration

### PR DESCRIPTION
**Pull request by trueqbit**
The original intention was to introduce pre-filtering of tuples before iteration, but this PR again contains much more again:

* Dedicated template metaprogramming sub-library that allows for complex, but readable filter conditions on traits
* Filter tuples prior to iteration
* Reduced combinatorial instantiation of generic lambdas for columns by taking `column_t`'s base classes as argument
* Removed superfluous if-statements in `prepared_statement_base`
* Simplified internal creation of indexed column
* Updated range insert/replace, utilizing `std::identity` projection (Replacing `default_transformer`)
* Replaced try/catch by `REQUIRE_THROWS_WITH` in unit tests
* Renamed `table_type` to `table_type_of` for disambiguation
* Moved 2 `storage_t` methods not dependent on templates to `storage_base`
* Simplified `iterator`
* Wrap calls to `sqlite3_prepare_v2()`
* Some containers can be directly initialized from a tuple
* Expand tuples directly into statement and row binders
* Removed explicit object+field type template parameters from `column_t`
* Split orthogonal information about a column into multiple base classes (this has been done already, however followed by column constraints).
* Started to use 'constrained algorithms' from the STL when possible
* Moved unused storage traits to unit tests
* Added appveyor build instructions for macOS (building is inactive)

### CODE REVIEW

File: dev/tuple_helper/tuple_traits.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+#include <tuple>
+
+#include "../functional/cxx_polyfill.h"
+#include "../functional/mpl.h"
+
+namespace sqlite_orm {
+    namespace internal {
+        /*
+         *  Higher-order trait metafunction that checks whether a tuple contains a type with given trait.
+         *
```
Comment (fnc12) [inline dev/tuple_helper/tuple_traits.h:13]: looks like a redundant line

File: dev/iterator.h [hunk 2]
```diff
@@ -24,15 +26,13 @@ namespace sqlite_orm {
...
           protected:
             /**
-             *  The double-indirection is so that copies of the iterator
-             *  share the same sqlite3_stmt from a sqlite3_prepare_v2()
-             *  call. When one finishes iterating it the pointer
-             *  inside the shared_ptr is nulled out in all copies.
+             *  shared_ptr is used over unique_ptr here
+             *  so that the iterator can be copyable.
              */
-            std::shared_ptr<statement_finalizer> stmt;
```
Comment (fnc12) [inline dev/iterator.h:32]: where finalization happens now?

File: dev/view.h [hunk 3]
```diff
@@ -39,11 +39,11 @@ namespace sqlite_orm {
             view_t(storage_type& stor, decltype(connection) conn, Args&&... args_) :
                 storage(stor), connection(std::move(conn)), args{std::make_tuple(std::forward<Args>(args_)...)} {}
 
-            size_t size() {
+            size_t size() const {
```
Comment (fnc12) [inline dev/view.h:42]: why did you add right `const`? right `const` means that nothing that is not mutable is changed. But `size` and `empty` calls access database so these calls can open a file and change opened files OS'es configuration.

File: dev/iterator.h (hunk 2 above)
Comment (trueqbit) [reply dev/iterator.h:32]: See next line ;)

File: dev/view.h (hunk 3 above)
Comment (trueqbit) [reply dev/view.h:42]: Sorry! I didn't recognize that it's accessing the database...

File: dev/tuple_helper/tuple_traits.h (hunk 1 above)
Comment (trueqbit) [reply dev/tuple_helper/tuple_traits.h:13]: set.

File: dev/table.h [hunk 4]
```diff
@@ -80,10 +84,6 @@ namespace sqlite_orm {
                 using F = member_field_type_t<M>;
                 const F* res = nullptr;
                 this->for_each_column_with_field_type<F>([&res, &memberPointer, &object](auto& column) {
-                    if(res) {
```
[the remark this answers is not in the document]
Comment (trueqbit) [reply dev/table.h:83]: Because `compare_any()` does the job.

File: dev/serializing_util.h [hunk 5]
```diff
@@ -213,7 +214,8 @@ namespace sqlite_orm {
 
             constexpr std::array<const char*, 2> sep = {", ", ""};
             for(size_t i = 0, first = true; i < args.size(); ++i) {
-                ss << sep[std::exchange(first, false)] << serialize(args[i], context);
+                ss << sep[std::exchange(first, false)]  ///
```
[the remark this answers is not in the document]
Comment (trueqbit) [reply dev/serializing_util.h:217]: It's a clang-format break. I saw it once in sqlite_orm, so I reused it.

File: dev/row_extractor.h [hunk 6]
```diff
@@ -276,77 +276,42 @@ namespace sqlite_orm {
     template<>
     struct row_extractor<std::vector<char>> {
         std::vector<char> extract(const char* row_value) const {
-            if(row_value) {
-                auto len = ::strlen(row_value);
-                return this->go(row_value, len);
-            } else {
-                return {};
-            }
+            return {row_value, row_value + (row_value ? ::strlen(row_value) : 0)};
```
[the remark this answers is not in the document]
Comment (trueqbit) [reply dev/row_extractor.h:279]: nothing, because `length` is 0, too. No memory will be accessed so passing a `nullptr` to the vector is ok.

File: dev/column_names_getter.h [hunk 7]
```diff
@@ -91,12 +91,7 @@ namespace sqlite_orm {
                 auto newContext = context;
                 newContext.skip_table_name = false;
                 iterate_tuple(cols.columns, [&columnNames, &newContext](auto& m) {
-                    auto columnName = serialize(m, newContext);
-                    if(!columnName.empty()) {
-                        columnNames.push_back(columnName);
-                    } else {
-                        throw std::system_error{orm_error_code::column_not_found};
-                    }
+                    columnNames.push_back(serialize(m, newContext));
```
[the remark this answers is not in the document]
Comment (trueqbit) [reply dev/column_names_getter.h:94]: The nested serialization takes care of that.

File: dev/serializing_util.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (fnc12) [reply dev/serializing_util.h:217]: what is the purpose of it?

[the remark this answers is not in the document]
Comment (trueqbit) [reply dev/serializing_util.h:217]: The purpose of the triple slash is to force a line-break. If you ask to remove the line break here, let me know :)

[the remark this answers is not in the document]
Comment (fnc12) [reply dev/serializing_util.h:217]: ok. I did not know that. About line break: why did you add line break here?

[the remark this answers is not in the document]
Comment (trueqbit) [reply dev/serializing_util.h:217]: Good question, I thought it made it a bit more readable and fit well with other places (in the same file) where 'tail streaming' cannot be used (i.e. the repeated use of `operator<<`).
What do you prefer?

File: dev/storage_base.h [hunk 8]
```diff
@@ -130,9 +150,9 @@ namespace sqlite_orm {
             }
 
             /**
-             *  Returns libsqltie3 lib version, not sqlite_orm
+             *  Returns libsqlite3 version, not sqlite_orm
              */
-            std::string libversion() {
+            static std::string libversion() {
```
[the remark this answers is not in the document]
Comment (trueqbit) [reply dev/storage_base.h:155]: I understand. The motivation behind this is simply that the library version is completely independent of everything else.
And you can still access the static method via the `object.member()` notation.

File: tests/prepared_statement_tests/get.cpp [hunk 9]
```diff
@@ -71,22 +72,13 @@ TEST_CASE("Prepared get") {
...
-                    REQUIRE(true);
-                }
+                REQUIRE_THROWS_WITH(storage.execute(statement), Contains("Not found"));
             }
         }
     }
     {
         auto statement = storage.prepare(get<User>(3));
         testSerializing(statement);
-        SECTION("nothing") {
```
[the remark this answers is not in the document]
Comment (trueqbit) [reply tests/prepared_statement_tests/get.cpp:87]: I understand what you are saying.
If you want to test a specific behaviour of the statement dtor, then we should do so explicitly, similar to what I implemented for the ["Transaction guard"](https://github.com/fnc12/sqlite_orm/blob/ef0695e40c93ba172e4416d29909cf1e574c7faa/tests/transaction_tests.cpp#L203).

File: dev/serializing_util.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (trueqbit) [reply dev/serializing_util.h]: Alright, so I removed all other `///` as well (except in unit tests).

File: dev/storage_base.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (fnc12) [reply dev/storage_base.h:155]: Yeah I know. Anyway I'd leave it untouched

File: tests/prepared_statement_tests/get.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (fnc12) [reply tests/prepared_statement_tests/get.cpp:87]: yes it is a good idea. But I don't see any important difference between these two ways. So my main thought is please leave it in the either way

File: dev/storage_traits.h [hunk 10]
```diff
@@ -42,227 +39,83 @@ namespace sqlite_orm {
             template<class S, class O>
             struct storage_columns_count : storage_columns_count_impl<storage_find_impl_t<S, O>> {};
```
Comment (trueqbit) [inline dev/storage_traits.h:40]: @fnc12 `storage_columns_count`, `storage_foreign_keys_count`, `storage_fk_references`, `storage_foreign_keys` are not used, only in unit tests. Should we move them to unit tests or even remove them completely?

**Review by trueqbit — COMMENTED** [review]
Most storage traits can be moved to unit tests.

File: dev/storage_traits.h (hunk 10 above)
Comment (fnc12) [reply dev/storage_traits.h:40]: let's move them to ut. We must not have code which is never used but it can be helpful so let's leave it in ut. If it is not required during next two versions then let's remove it

File: dev/storage_traits.h [hunk 11]
```diff
@@ -1,24 +1,18 @@
...
-        template<class... Ts>
-        struct storage_impl;
-
-        template<class T, bool WithoutRowId, class... Cs>
-        struct table_t;
-
-        template<class A, class B>
-        struct foreign_key_t;
-
         namespace storage_traits {
```
Comment (trueqbit) [inline dev/storage_traits.h:16]: @fnc12 Do you want to keep the `storage_traits` nested namespace?

File: dev/storage_traits.h (hunk 10 above)
Comment (trueqbit) [reply dev/storage_traits.h]: Do you want to keep the `storage_traits` nested namespace?

File: dev/storage_traits.h (hunk 11 above)
Comment (fnc12) [reply dev/storage_traits.h:16]: yes

File: dev/storage_traits.h (hunk 10 above)
Comment (fnc12) [reply dev/storage_traits.h]: yes

**Comment by fnc12** [comment]
thanks @trueqbit
